### DISCUSSION TITLE
How to insert when conditions are met, or not?

### DISCUSSION THREAD

**Discussion by ddxl123** [simolus3/drift#2204] [Q&A]
How to insert when conditions are met, or not?

For example, if there is a field `name` in a table, when I insert a new row, it will be inserted only if the `name`'s value of the new row does not exist, and it will not be inserted if it exists. The `name` field is not set as a unique constraint.

Without using query statements.

**Comment by simolus3** [comment]
If you really must avoid compiled queries and can't set a unique constraint on `name`, I think the most intuitive way to approach this is to just send two statements in a transaction for atomicity.

So you'd first run a select to see if any row already exists - if there isn't, you run the insert.

```dart
  Future<void> insertIfNameDoesNotExist(Insertable<TodoItem> item) async {
    await transaction(() async {
      final asColumns = item.toColumns(true);
      final matchingName =
          asColumns['title']!; // change this with the "name" column
      final withSameName = todoItems.title.count();
      final withSameNameQuery = selectOnly(todoItems)
        ..addColumns([withSameName])
        // Same here, .title needs to be changed with the name of the column
        ..where(todoItems.title.equalsExp(matchingName.dartCast()));

      final count = (await withSameNameQuery.getSingle()).read(withSameName)!;
      if (count == 0) {
        await into(todoItems).insert(item);
      }
    });
  }
```
